### PR TITLE
Fix concurrent writes in the update operation

### PR DESCRIPTION
**Pull request by ospfranco**
### Details

The update operation was written in such manner that it fired all database operations at once. Some of the operations internally rely on other promises, meaning they will get re-scheduled on another tick. This causes unpredictable behavior as you cannot guarantee the order of the operations. The solution is clear and is the only one, all DB writes must respect the order of the operations.

The incorrect ordering of the operations being committed to the DB causes other random behavior around the app as well. Linked the two issues.

### Related Issues

https://github.com/Expensify/App/issues/28737
https://github.com/Expensify/App/pull/28584

### Automated Tests

Added one more test that makes sure the correct data is persisted. It failed with the previous implementation, now it passes.

### CODE REVIEW

File: tests/unit/onyxTest.js [hunk 1]
```diff
@@ -899,9 +899,9 @@ describe('Onyx', () => {
             {onyxMethod: Onyx.METHOD.MERGE_COLLECTION, key: ONYX_KEYS.COLLECTION.TEST_UPDATE, value: {[itemKey]: {a: 'a'}}},
         ])
             .then(() => {
-                expect(collectionCallback).toHaveBeenNthCalledWith(1, {[itemKey]: {a: 'a'}});
-                expect(testCallback).toHaveBeenNthCalledWith(1, 'taco', ONYX_KEYS.TEST_KEY);
-                expect(otherTestCallback).toHaveBeenNthCalledWith(1, 'pizza', ONYX_KEYS.OTHER_TEST);
+                expect(collectionCallback).toHaveBeenNthCalledWith(2, {[itemKey]: {a: 'a'}});
+                expect(testCallback).toHaveBeenNthCalledWith(2, 'taco', ONYX_KEYS.TEST_KEY);
+                expect(otherTestCallback).toHaveBeenNthCalledWith(2, 'pizza', ONYX_KEYS.OTHER_TEST);
```
Comment (ospfranco) [inline tests/unit/onyxTest.js:904]: This previously was receiving the write updates before the hydration callback, but now that the write operations take a little longer the order is correct.

File: lib/Onyx.js [hunk 2]
```diff
@@ -1414,32 +1436,7 @@ function update(data) {
...
-            case METHOD.CLEAR:
-                clearPromise = clear();
-                break;
-            default:
-                break;
-        }
-    });
-
-    return clearPromise.then(() => Promise.all(_.map(promises, p => p())));
+    return innerUpdate(data);
```
Comment (ospfranco) [inline lib/Onyx.js:1439]: As a data point, [I tried the proposed solution by Marc](https://github.com/Expensify/react-native-onyx/compare/marcaaron-resolveOnBroadcast) which would try to resolve earlier while keeping the writes async, this however is not correct and broke a bunch of other tests.

**Review by tgolen — CHANGES_REQUESTED** [review]
Has there been any testing done in App to see what kind of performance impact this will have? If so, can you please attach it to this PR?

File: lib/Onyx.js [hunk 3]
```diff
@@ -1534,6 +1539,7 @@ const Onyx = {
     connect,
     disconnect,
     set,
+    get,
```
[the remark this answers is not in the document]
Comment (ospfranco) [reply lib/Onyx.js:1542]: From what I saw Onyx.connect has caching implemented, which might not be what is actually persisted on the DB. I was about to directly query the DB which is at the end of the day what really matters. I think this `get` might also use caches right? if so, checking the DB should be the correct path.

**Comment by ospfranco** [comment]
No, I haven't tested the app. I can try to get some metrics but I already asked in slack if there is any particular call to update that is heavy.

File: lib/Onyx.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (tgolen) [reply lib/Onyx.js:1542]: I like the idea of querying the DB directly if that's possible.

**Comment by ospfranco** [comment]
I added some react-native-performance measures on app start, I cannot tell which one exactly is OpenApp but they are all basically negible. Writing to DB should not be the bottleneck, and even then, correctness is not negotiable.

Oscar Franco Screen 000065

File: lib/Onyx.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (ospfranco) [reply lib/Onyx.js]: Done

**Comment by tgolen** [comment]
OK, I just created a large account for testing things like this, so let me give it a try with this today.

File: lib/Onyx.js [hunk 4]
```diff
@@ -1392,6 +1392,36 @@ function mergeCollection(collectionKey, collection) {
         });
 }
 
+function innerUpdate(data) {
```
Comment (tgolen) [inline lib/Onyx.js:1395]: Please add JSDoc

**Review by tgolen — CHANGES_REQUESTED** [review]
I tested it in App with a large account and I couldn't notice any performance impact, so I think these are good changes. I just noticed this one thing.

**Comment by marcaaron** [comment]
Doing some testing now...

**Comment by marcaaron** [comment]
Alright well - it is defying my expectations a bit, but when timing the calls to `Onyx.update()` they are running way faster synchronously 😅 

### Sign in trials

**With sync updates:**

1

Timing:development.new.expensify.flushQueue 458
Timing.js:49 Timing:development.new.expensify.flushQueue 0
Timing.js:49 Timing:development.new.expensify.flushQueue 407

2

Timing:development.new.expensify.flushQueue 468
Timing.js:49 Timing:development.new.expensify.flushQueue 1
Timing.js:49 Timing:development.new.expensify.flushQueue 191

3 

Timing:development.new.expensify.flushQueue 311
Timing.js:49 Timing:development.new.expensify.flushQueue 0
Timing.js:49 Timing:development.new.expensify.flushQueue 226

**Without sync updates (but on latest `react-native-onyx`):**

1

Timing:development.new.expensify.flushQueue 488
Timing.js:49 Timing:development.new.expensify.flushQueue 4147
Timing.js:49 Timing:development.new.expensify.flushQueue 962

2

Timing:development.new.expensify.flushQueue 420
Timing.js:49 Timing:development.new.expensify.flushQueue 1126
Timing.js:49 Timing:development.new.expensify.flushQueue 170

3

Timing:development.new.expensify.flushQueue 464
Timing.js:49 Timing:development.new.expensify.flushQueue 4729
Timing.js:49 Timing:development.new.expensify.flushQueue 976

This was a web-only test. Haven't tested on iOS or Android - but looks promising.

File: lib/Onyx.js [hunk 5]
```diff
@@ -1414,32 +1444,7 @@ function update(data) {
...
-                break;
-            case METHOD.CLEAR:
-                clearPromise = clear();
-                break;
-            default:
-                break;
-        }
-    });
-
-    return clearPromise.then(() => Promise.all(_.map(promises, p => p())));
```
Comment (marcaaron) [inline lib/Onyx.js:1442]: I see we are undoing the previous logic to move the `clear()` to the front of the update queue. Does running the updates sync solve the issue that caused us to make that change?

Comment (ospfranco) [reply lib/Onyx.js:1442]: ah, probably not. I'll take care of it.

Comment (ospfranco) [reply lib/Onyx.js:1442]: I'm not sure making sure clear runs first is a desired behavior but I guess now it is too late to change it, other parts of the system might rely on this behavior. I added sorting to the list to make sure clear operations get executed first.

**Review by deetergp — APPROVED** [review]

File: lib/Onyx.js (hunk 5 above)
Comment (marcaaron) [reply lib/Onyx.js:1442]: Yeah, I agree it is a bit of a weird hack. But changing it could have unexpected consequences. Thanks for updating.

**Review by marcaaron — APPROVED** [review]
🚀

File: lib/Onyx.js [hunk 6]
```diff
@@ -1392,6 +1392,42 @@ function mergeCollection(collectionKey, collection) {
...
+            break;
+        case METHOD.MULTI_SET:
+            promise = multiSet(value);
+            break;
+        case METHOD.CLEAR:
+            promise = clear();
+            break;
+        default:
+            break;
+    }
```
Comment (tgolen) [inline lib/Onyx.js:1426]: It looks like this change possibly led to some regressions spotted in [this PR](https://github.com/Expensify/App/pull/29169#issuecomment-1755269426) and discussed in [this Slack thread](https://expensify.slack.com/archives/C035J5C9FAP/p1696941880028159?thread_ts=1696934662.824599&cid=C035J5C9FAP) that updated the Onyx version

**Pull request merged**